sunpy/io/jp2.py

```diff
@@ -72,5 +74,107 @@ def get_header(filepath):
     return [FileHeader(pydict)]
 
 
+def header_to_xml(header):
+    """
+    Converts image header metadata into an XML Tree that can be inserted into
+    a JP2 file header.
+
+    Parameters
+    ----------
+    header : `MetaDict`
+        A header dictionary to convert to xml.
+
+    Returns
+    ----------
+    `lxml.etree._Element`
+        A fits element where each child is an xml element
+        in the form <key>value</key> derived from the key/value
+        pairs in the given header dictionary
+    """
+    # glymur uses lxml and will crash if trying to use
+    # python's builtin xml.etree
+    import lxml.etree as ET
+
+    fits = ET.Element("fits")
+
+    already_added = set()
+    for key in header:
+        # Some headers span multiple lines and get duplicated as keys
+        # header.get will appropriately return all data, so if we see
+        # a key again, we can assume it was already added to the xml tree.
+        if (key in already_added):
+            continue
+
+        # Add to the set so we don't duplicate entries
+        already_added.add(key)
+
+        el = ET.SubElement(fits, key)
+        data = header.get(key)
+        if type(data) == bool:
+            data = "1" if data else "0"
+        else:
+            data = str(data)
+
+        el.text = data
+
+    return fits
+
+
+def generate_jp2_xmlbox(header):
+    """
+    Generates the JP2 XML box to be inserted into the jp2 file.
+
+    Parameters
+    ----------
+    header : `MetaDict`
+        A header dictionary.
+
+    Returns
+    ----------
+    `XMLBox`
+        XML box containing FITS metadata to be used in jp2 headers
+    """
+    # glymur uses lxml and will crash if trying to use
+    # python's builtin xml.etree
+    import lxml.etree as ET
+    from glymur import jp2box
+
+    header_xml = header_to_xml(header)
+    meta = ET.Element("meta")
+    meta.append(header_xml)
+    tree = ET.ElementTree(meta)
+    return jp2box.XMLBox(xml=tree)
+
+
 def write(fname, data, header):
-    raise NotImplementedError("No jp2 writer is implemented.")
+    """
+    Take a data header pair and write a JP2 file.
+
+    Parameters
+    ----------
+    fname : `str`
+        File name, with extension.
+    data : `numpy.ndarray`
+        n-dimensional data array.
+    header : `dict`
+        A header dictionary.
+    """
+    from glymur import Jp2k
+
+    # Create an initial jp2 file with the given data
+    tmpname = fname + "tmp.jp2"
+    jp2_data = np.uint8(data)
+    jp2 = Jp2k(tmpname, jp2_data)
+
+    # Append the XML data to the headerp information
+    meta_boxes = jp2.box
+    target_index = len(meta_boxes) - 1
+    fits_box = generate_jp2_xmlbox(header)
+    meta_boxes.insert(target_index, fits_box)
+
+    # Rewrite the jp2 file with the xml data in the header
+    jp2.wrap(fname, boxes=meta_boxes)
+
+    # Remove the initial temporary file
```

Comment:
```suggestion
```

Reply:
Why do you want to remove the comments?

Reply:
I don't think they are useful.

Reply:
They describe the process that's being done rather than trying to figure out what's going on inside glymur.
Like the fact that Jp2k and jp2.wrap actually write the file.
And that the these arbitrary "boxes" are part of the file's header.

Reply:
Then the middle two comments could stay?

Reply:
Sure

Reply:
Updated